src/Microsoft.AspNet.Mvc.Core/Logging/DefaultActionSelectorLoggerExtensions.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.Mvc.ActionConstraints;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Mvc.Logging
+{
+    public static class DefaultActionSelectorLoggerExtensions
+    {
+        private static Action<ILogger, string, Exception> _ambiguousActions;
+        private static Action<ILogger, string, string, IActionConstraint, Exception> _constraintMismatch;
+
+        static DefaultActionSelectorLoggerExtensions()
+        {
+            _ambiguousActions = LoggerMessage.Define<string>(
+                LogLevel.Error,
+                1,
+                "Request matched multiple actions resulting in ambiguity. " +
+                    "Matching actions: {AmbiguousActions}");
+            _constraintMismatch = LoggerMessage.Define<string, string, IActionConstraint>(
+                LogLevel.Verbose,
+                2,
+                "Action '{ActionName}' with id '{ActionId}' did not match the " +
+                    "constraint '{ActionConstraint}'");
+        }
+
+        public static void AmbiguousActions(this ILogger logger, string actionNames)
```

Comment:
When we had `[NotNull]` we'd typically decorate the heck out of all of our public extension methods. Now that we don't, with this new pattern are we not doing manual `null` checks? I'm conflicted about it myself which is why i ask because it'd add a to of boiler plate. RIP `[NotNull]`